src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/abs/TestRouter.java

```diff
@@ -0,0 +1,172 @@
+     * Constructor for this class.
+     *
+     * @param executor An {@link ExecutorService} to run tasks asynchronously.
+     * @param dbPort   The real DB server port.
+     * @throws IOException if some problem occurs creating a new {@link ServerSocket}.
+     */
+    TestRouter(final ExecutorService executor, final int dbPort) throws IOException {
+        this.executor = executor;
+        this.dbPort = dbPort;
+    }
+
+    /**
+     * Gets new DB properties with the JDBC URL patched to replace the default DB port with a new one.
+     *
+     * @param dbProperties The original DB properties.
+     * @param defaultPort  The default DB port.
+     * @param newPort      The port to use as replacement in the JDBC URL for the default DB port.
+     * @return The patched DB properties.
+     */
+    public static Properties getPatchedDbProperties(final Properties dbProperties, final int defaultPort, final int newPort) {
+        final Properties testProps = new Properties();
+        testProps.putAll(dbProperties);
+
+        final String patchedJdbc = dbProperties.getProperty(JDBC)
+                .replace("localhost:" + defaultPort, "localhost:" + newPort);
+        testProps.setProperty(JDBC, patchedJdbc);
+
+        return testProps;
+    }
+
+    /**
+     * Gets the local port (that will act as the DB server port for the DB engine).
+     *
+     * @return the local port.
+     */
+    public int getLocalPort() {
+        return serverSocket.getLocalPort();
+    }
+
+    /**
+     * Gets the database port (the port where the DB server being tested is listening).
+     *
+     * @return the database port.
+     */
+    public int getDbPort() {
+        return dbPort;
+    }
+
+    /**
+     * Initializes connections to/from the DB server.
+     *
+     * @throws IOException if there is a problem establishing connections.
+     */
+    public void init() throws IOException {
+        socketRouterToDb.connect(new InetSocketAddress(InetAddress.getLoopbackAddress(), dbPort));
+
+        final Future<Socket> socketFuture = executor.submit(serverSocket::accept);
+
+        executor.submit(() -> {
+                    final Socket socketTestToRouter = socketFuture.get();
+                    logger.info("Test-to-DB established");
+                    while (isRunning) {
+                        int read = socketTestToRouter.getInputStream().read();
+                        socketRouterToDb.getOutputStream().write(read);
+                    }
+                    return null;
+                }
+        );
+
+        executor.submit(() -> {
+                    final Socket socketTestToDb = socketFuture.get();
+                    logger.info("DB-to-Test established");
+                    while (isRunning) {
+                        int read = socketRouterToDb.getInputStream().read();
+                        socketTestToDb.getOutputStream().write(read);
+                    }
+                    return null;
+                }
+        );
```

Comment:
For this to work smoothly you would need an executor with at least 3 threads. But the executor is passed as an argument to the constructor, so someone could by mistake use an executor with only 1 or 2 threads.

IMO this executor should be internal to this class and instatiated in the constructor.